src/app/data-model/BUILD.gn

```diff
@@ -0,0 +1,57 @@
+# Copyright (c) 2024 Project CHIP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import("//build_overrides/chip.gni")
+
+source_set("data-model") {
+  sources = [
+    "BasicTypes.h",
+    "DecodableList.h",
+    "Decode.h",
+    "Encode.h",
+    "FabricScoped.h",
+    "List.h",
+    "NullObject.h",
+    "Nullable.h",
+    "TagBoundEncoder.h",
+    "WrappedStructEncoder.h",
+  ]
+
+  deps = [
+    # TODO: dependencies NOT declared because they are not part of
+    #       any GN dependency. Overall src/app seems to suffer greatly
+    #       of this, in part due to zap-generated code dependency.
+    #
+    #  - app/util/attribute-storage-null-handling.h
+    #  - app/ConcreteAttributePath.h
```

Comment:
Hmm.  Those are not part of any dependency just because they were added by someone who doesn't understand gn (me, at the time).

Followup for this part is great; don't try to add it to this PR, but we should be able to have some sort of source_set or set of source_sets for "src/app stuff that does not depend on icky things", which would include both of those headers.

Reply:
That is my plan: I am adding things piecewise, however we have 81 orphaned files in src/app and I cannot add all dependencies in one (and some are really hard, like all ember depending on generated files).

End goal is all to be clearly defined, however those will take several PRs that each incrementally add a few more "known files" into gn files.